libraries/ESP8266WiFiMesh/src/MeshBackendBase.h

```diff
@@ -0,0 +1,315 @@
+/*
+  MeshBackendBase
+ 
+  Copyright (c) 2015 Julian Fell and 2019 Anders Löfgren. All rights reserved.
+ 
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#ifndef __MESHBACKENDBASE_H__
+#define __MESHBACKENDBASE_H__
+
+#include <ESP8266WiFi.h>
+#include "TransmissionOutcome.h"
+#include "NetworkInfoBase.h"
+
+const String ESP8266_MESH_EMPTY_STRING = "";
+
+typedef enum
+{
+  MB_TCP_IP = 0,
+  MB_ESP_NOW = 1
+} mesh_backend_t;
+
+class MeshBackendBase {
```

Comment:
For archive purposes: I'm not sure this is possible. I've tried to design the library in such a way that both subclasses (`TcpIpMeshBackend` and `EspnowMeshBackend`) can be used in the same node at the same time, so that the same node can use both ESP-NOW and TCP/IP simultaneously. As a result of this, some aspects of the backend will change dynamically during runtime. One example is the variable `MeshBackendBase *apController` which contains whichever backend instance is currently in charge of the AP. This can change depending on the input from the outside world to the node, so there is no way to know it at compile-time. Since the subclasses also differ greatly in available methods, there are many cases where they cannot substitute each other.